File: 1339-Maximum-Product-of-Splitted-Binary-Tree/1339-Maximum-Product-of-Splitted-Binary-Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxProduct(self, root: Optional[TreeNode]) -> int:
        ans = []
        def find(root):
            #base condition for a leaf node
            if not root: return 0
            sum = find(root.left) + find(root.right) + root.val
            #adding the possible sum of subtree into the stack ans
            ans.append(sum)
            return sum
        totalsum = find(root)
        #returning the max of products of sum of 2 subtress
        return int(max((totalsum-sum)*sum for sum in ans) % (1e9+7))
      
      #TC: O(n)
      #SC: O(h) 

File: 1339-Maximum-Product-of-Splitted-Binary-Tree/1339-Maximum-Product-of-Splitted-Binary-Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    long long totalsum = 0, result = 0;
    long long int MOD = 1e9+7;
    int find(TreeNode* root){
        if (root == NULL){
            return 0;
        }
        long long sum = find(root->left) + find(root->right) + root->val;
        result = max(result, (totalsum-sum)*sum);
        return sum;
    }
    
    int maxProduct(TreeNode* root) {
        totalsum = find(root);
        find(root);
        return result % MOD;
    }
};
